tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+  httr::stop_for_status(resp) # covers outcomes 1 *and* 3 from above list
+  if (httr::content(resp)$responseCode == 200) {
+    # covers outcome 2b
+    stop("The handle exists, but the queried value(s) could not be found.")
+  }
+  # this must be outcome 2a
+  resp
+}
+
+# still missing: a way to disambiguate between outcomes 1 and 3
+# exported function, this is tested
+is_doi_found <- function(doi) {
+  # HEAD is more idiomatic, and marginally more efficient
+  resp <- verb_doi_handle(doi = doi, verb = "HEAD")
+  if (httr::status_code(resp) == 200) return(TRUE)
+  if (httr::status_code(resp) == 404) return(FALSE)
+  httr::stop_for_status(resp)
+}
+```
+
+A minimal interface between exported functions and the API is desirable, because it simplifies the below tests.
+`get_doi_handle()` would be the absolute minimum, but it alone cannot distinguish between API failure modes 1 from 3:
+`httr::stop_for_status()` could be thrown either because a DOI does not exist or because of some other HTTP error code.
+The differences between 1 and 3 are important for users and the downstream logic in the biblids package.
+We thus need an additional function at this interface, `is_doi_found()`, which checks explicity for failure mode 1.
+
+We can test most API failure modes of `get_doi_handle()` and `is_doi_found()` thus:
+
+```{r test-minimal-wrapper}
+library(testthat)
+skip_if_offline()
+# outcome 1
+test_that("Non-existent DOIs error out", {
+  expect_error(get_doi_handle("10.1000/I_DO_NOT_EXIST"))
+  expect_false(is_doi_found("10.1000/I_DO_NOT_EXIST"))
+})
+# outcome 2a
+test_that("Existing DOIs are found", {
+  testthat::local_edition(3)
+  # this snapshot will also capture some change in the API output
+  expect_snapshot_output(get_doi_handle("10.1038/nature14236"))
+  expect_true(is_doi_found("10.1038/nature14236"))
+})
+# outcome 2b
+test_that("Non-existing values error out", {
+  expect_error(
+    get_doi_handle(
+      doi = "10.1038/nature14236",
+      query = list(type = "I_DO_NOT_EXIST")
+    )
+  )
+})
+```
+
+One might think that these tests should be mocked, rather than running against the live API.
+However, because DOI.org is stable, performant and requires no authentication, the added complexity is not worth it in this case.
+Moreover, `get_doi_handle()` and `is_doi_found()` perform minimal work and are essentially I/O functions.
+They may thus require *only* (broad) integration testing against a live API.
+
+We do, however, need a mock to test the remaining outcome 3, which cannot readily be retrieved from the live API.
+
+```{r mock-minimal}
+library(httptest)
+test_that("Other http error codes are captured", {
+  expect_error(without_internet(is_doi_found("10.1038/nature14236")))
+  expect_error(without_internet(get_doi_handle("10.1038/nature14236")))
+})
+```
+
+Happily, we can cheat a little, and simply create *some* sure HTTP status error, without actually reproducing API behavior in any detail.
```

Comment:
the whole book is about happily cheating :grin:

Reply:
I think it's very rigorous 😊